datajunction-server/datajunction_server/sql/decompose.py

```diff
@@ -0,0 +1,146 @@
+"""Used for extracting measures form metric definitions."""
+from pydantic import BaseModel
+
+from datajunction_server.enum import StrEnum
+from datajunction_server.sql import functions as dj_functions
+from datajunction_server.sql.parsing.backends.antlr4 import ast, parse
+
+
+class Aggregability(StrEnum):
+    """
+    Type of allowed aggregation for a given measure.
+    """
+
+    FULL = "full"
+    LIMITED = "limited"
+    NONE = "none"
+
+
+class AggregationRule(BaseModel):
+    """
+    Measures are components used to build metrics (e.g., sales_amount, revenue, user_count).
+    """
+
+    type: Aggregability = Aggregability.NONE
+    level: list[str] | None = None
+
+
+class Measure(BaseModel):
+    """
+    Measures are components used to build metrics (e.g., sales_amount, revenue, user_count).
+    """
+
+    name: str
+    expression: str  # A SQL expression for defining the measure
+    aggregation: str
+    rule: AggregationRule
+
+
+class MeasureExtractor:
+    """
+    Extracts aggregatable measures from a metric definition and generates SQL
+    derived from those measures.
+    """
+
+    def __init__(self):
+        """Register handlers for aggregation functions"""
+        self.handlers = {
+            dj_functions.Sum: self._generic_additive_agg,
+            dj_functions.Count: self._generic_additive_agg,
+            dj_functions.Max: self._generic_additive_agg,
+            dj_functions.Min: self._generic_additive_agg,
+            dj_functions.Avg: self._avg,
+        }
+
+    def extract_measures(self, metric_query: str) -> tuple[list[Measure], ast.Query]:
+        """
+        Decomposes the metric query into its constituent aggregatable measures and
+        constructs a SQL query derived from those measures.
+        """
+        query_ast = parse(metric_query)
+        measures = []
+
+        for idx, func in enumerate(query_ast.find_all(ast.Function)):
+            handler = self.handlers.get(func.function())
+            if handler:
+                func_measures = handler(func, idx)
+                if func_measures:
+                    self.update_ast(func, func_measures)
+                measures.extend(func_measures)
+
+        return measures, query_ast
+
+    def _generic_additive_agg(self, func, idx) -> list[Measure]:
+        arg = func.args[0]
+        measure_name = "_".join(
+            [str(col) for col in arg.find_all(ast.Column)] + [func.name.name.lower()],
```

Comment:
In general no, it's only meant to support aggregation functions like `SUM(a)`, `MAX(b)` etc. I wouldn't expect them to take multiple arguments. Maybe a better name for this function is `_simple_associative_agg`.

Reply:
If we don't expect to handle multiple arguments, do we need/want `arg.find_all(ast.Column)`? Should we either find the first Column only, or raise when multiple are found?

Reply:
Actually, we're already only picking the first argument to the aggregation function - see line 74: `arg = func.args[0]`. However, we still need `arg.find_all(ast.Column)`, because that will find *all* referenced columns in the argument, which might not be just a single column.

For example, if you have `SUM(view_secs + clicks)`, there is a single argument to`SUM` which is `view_secs + clicks`, and `arg.find_all(ast.Column)` will return both of the columns referenced in that single argument: `[ast.Column("view_secs"), ast.Column("clicks")]`. We're then using the name of both columns to construct the measure name. 

Alternatively, we could name the measures something much less intuitive (e.g., just increase a counter and have `m0`, `m1`, `m2` etc), but I figured that actually naming them something sensible would be helpful to users parsing the metadata.

Reply:
Makes sense! Thank you for explaining!